proxyd/consensus_aware.go

```diff
@@ -0,0 +1,392 @@
+
+	lastUpdate time.Time
+
+	bannedUntil time.Time
+}
+
+type ConsensusPoller struct {
+	cancelFunc context.CancelFunc
+
+	BackendGroup      *BackendGroup
+	backendState      map[*Backend]*backendState
+	consensusGroupMux sync.Mutex
+	ConsensusGroup    []*Backend
+
+	Tracker      ConsensusTracker
+	asyncHandler ConsensusAsyncHandler
+}
+
+type ConsensusTracker interface {
+	ConsensusBlockNumber() string
+	SetConsensusBlockNumber(blockNumber string)
+}
+
+type InMemoryConsensusTracker struct {
+	consensusBlockNumber string
+	mutex                sync.Mutex
+}
+
+func (ct *InMemoryConsensusTracker) ConsensusBlockNumber() string {
+	defer ct.mutex.Unlock()
+	ct.mutex.Lock()
+
+	return ct.consensusBlockNumber
+}
+
+func (ct *InMemoryConsensusTracker) SetConsensusBlockNumber(blockNumber string) {
+	defer ct.mutex.Unlock()
+	ct.mutex.Lock()
+
+	ct.consensusBlockNumber = blockNumber
+}
+
+func (cp *ConsensusPoller) GetConsensusGroup() []*Backend {
+	defer cp.consensusGroupMux.Unlock()
+	cp.consensusGroupMux.Lock()
+
+	g := make([]*Backend, 0, len(cp.BackendGroup.Backends))
+	copy(g, cp.ConsensusGroup)
+
+	return g
+}
+
+func (cp *ConsensusPoller) Shutdown() {
+	cp.asyncHandler.Shutdown()
+}
+
+func NewInMemoryConsensusTracker() ConsensusTracker {
+	return &InMemoryConsensusTracker{
+		consensusBlockNumber: "", // empty string semantics means unknown
+		mutex:                sync.Mutex{},
+	}
+}
+
+type RedisConsensusTracker struct {
+	ctx          context.Context
+	client       *redis.Client
+	backendGroup string
+}
+
+func (ct *RedisConsensusTracker) key() string {
+	return fmt.Sprintf("consensus_latest_block:%s", ct.backendGroup)
+}
+
+func (ct *RedisConsensusTracker) ConsensusBlockNumber() string {
+	return ct.client.Get(ct.ctx, ct.key()).Val()
+}
+
+func (ct *RedisConsensusTracker) SetConsensusBlockNumber(blockNumber string) {
+	ct.client.Set(ct.ctx, ct.key(), blockNumber, 0)
+}
+
+func NewRedisConsensusTracker(ctx context.Context, r *redis.Client, namespace string) ConsensusTracker {
+	return &RedisConsensusTracker{
+		ctx:          ctx,
+		client:       r,
+		backendGroup: namespace,
+	}
+}
+
+// ConsensusAsyncHandler controls the asynchronous polling mechanism, interval and shutdown
+type ConsensusAsyncHandler interface {
+	Init()
+	Shutdown()
+}
+
+type NoopAsyncHandler struct{}
+
+func NewNoopAsyncHandler() ConsensusAsyncHandler {
+	log.Warn("using NewNoopAsyncHandler")
+	return &NoopAsyncHandler{}
+}
+func (ah *NoopAsyncHandler) Init()     {}
+func (ah *NoopAsyncHandler) Shutdown() {}
+
+type PollerAsyncHandler struct {
+	ctx context.Context
+	cp  *ConsensusPoller
+}
+
+func NewPollerAsyncHandler(ctx context.Context, cp *ConsensusPoller) ConsensusAsyncHandler {
+	return &PollerAsyncHandler{
+		ctx: ctx,
+		cp:  cp,
+	}
+}
+func (ah *PollerAsyncHandler) Init() {
+	// create the individual backend pollers
+	for _, be := range ah.cp.BackendGroup.Backends {
+		go func(be *Backend) {
+			for {
+				timer := time.NewTimer(PollerInterval)
```

Comment:
When we set the interval to, let's say 2 seconds, we expect the poller to refresh it's state 5 times in 10 seconds.

In the current impl if `UpdateBackend` takes more than 2 seconds we won't wait any longer to try to refresh the state. Therefore reaching the closer as possible to 5 in 10 seconds.

If we make the interval static in-between calls and the calls are taking 2 second themselves, it would add an extra 2 seconds interval, reaching at max 2.5 executions in 10 seconds.

I'm inclined to the first alternative, but let me know if you think the latter makes more sense.

Reply:
Resolved, current solution makes sense.